packages/flutter/lib/src/material/material_state.dart

```diff
@@ -760,6 +760,10 @@ class MaterialStatePropertyAll<T> implements MaterialStateProperty<T> {
 /// [MaterialState.focused] to its controller. When the widget gains the
 /// or loses the focus it will [update] its controller's [value] and
 /// notify listeners of the change.
+///
+/// When calling `setState` in a [MaterialStatesController] listener, use the
```

Comment:
Two issues here:

- When the controller's value is updated in a event/gesture handler or in didUpdateWidget, then it is safe to call setState. Sometimes the value is updated in a build method. Like SegmentedButton - `controller.update(MaterialState.selected, segmentSelected)` - and then it's not safe. So you could say that it's generally prudent to use a postFrameCallback, but the situation is a little more complicated then that.
- addPostFrameCallback defers the setState call until after the entire build/layout/paint cycle has finished. This comment makes it sound like it's only deferred until after MaterialStatesController.update returns.